platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +88,186 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+        Objects.requireNonNull(rate).getSellAmount(), sellAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'sell_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    // buy_amount has a proper number of significant decimals
+    if (!NumberHelper.hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getBuyAmount(), buyAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          "'buy_amount' has incorrect number of significant decimals in the GET /rate response",
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is set, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
+      if (fee.getAsset() == null || feeAsset == null) {
+        logErrorAndThrow(
+            "'fee.asset' is missing or not a valid asset in the GET /rate response",
+            ServerErrorException.class);
+      }
+
+      if (fee.getAsset().equals(request.getSellAsset())) {
+        // when fee is in sell_asset,
+        // check that sell_amount is equal to price * buy_amount + (fee ?: 0)
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()))
+                .add(new BigDecimal(fee.getTotal()));
+
+        // Since we don't know how the anchor rounds the amounts, we need to check the equality in
+        // all allowed rounding modes
+        if (!equalsInScale(
+            new BigDecimal(rate.getSellAmount()), expected, sellAsset.getSignificantDecimals())) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * buy_amount + (fee?:0) in the GET /rate response",
+              ServerErrorException.class);
+        }
+      } else {
+        // when fee is in buy_asset,
+        // check that sell_amount is equal to price * (buy_amount + (fee ?: 0))
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()).add(new BigDecimal(fee.getTotal())));
+        // Since we don't know how the anchor rounds the amounts, we need to check the equality in
+        // all allowed rounding modes
+        if (!equalsInScale(
+            new BigDecimal(rate.getSellAmount()), expected, sellAsset.getSignificantDecimals())) {
+          logErrorAndThrow(
+              "'sell_amount' is not equal to price * (buy_amount + (fee ?: 0)) in the GET /rate response",
+              ServerErrorException.class);
+        }
```

Comment:
This can be moved out of if/else branch

Reply:
They are actually different and cannot be moved out.
```
// check that sell_amount is equal to price * buy_amount + (fee ?: 0)
// check that sell_amount is equal to price * (buy_amount + (fee ?: 0))
```